File: solutions/go/src/playground/algorithm/combinations.go
package algorithm

func combine(n int, k int) [][]int {
	result := make([][]int, 0)
	combineR(1, n, 0, k, make([]int, k), &result)
	return result
}

func combineR(start, end int, depth, length int, combine []int, result *[][]int) {
	if depth >= length {
		temp := make([]int, length)
		copy(temp, combine)
		*result = append(*result, temp)
		return
	}

	for n := start; n <= end; n++ {
		combine[depth] = n
		combineR(n+1, end, depth+1, length, combine, result)
	}
}

File: solutions/java/src/main/java/playground/algorithm/Combinations.java
package playground.algorithm;

import java.util.ArrayList;
import java.util.List;

public class Combinations {

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        combine(1, n, 0, k, new ArrayList<>(), result);
        return result;
    }

    private void combine(int start, int end, int depth, int length, List<Integer> list, List<List<Integer>> result) {
        if (depth == length) {
            result.add(new ArrayList<>(list));
            return;
        }

        for (int i = start; i <= end; i++) {
            list.add(i);
            combine(i + 1, end, depth + 1, length, list, result);
            list.remove(list.size() - 1);
        }
    }
}
